@xen-orchestra/fs/src/azure.js

```diff
@@ -0,0 +1,334 @@
+import { BlobServiceClient, StorageSharedKeyCredential } from '@azure/storage-blob'
+import { createLogger } from '@xen-orchestra/log'
+import { parse } from 'xo-remote-parser'
+import { join, split } from './path'
+import RemoteHandlerAbstract from './abstract'
+import { pRetry } from 'promise-toolbox'
+import { readChunk } from '@vates/read-chunk'
+import copyStreamToBuffer from './_copyStreamToBuffer'
+
+createLogger('xo:fs:azure')
+const MAX_BLOCK_SIZE = 1024 * 1024 * 4000 // 4000 MiB
+const MIN_BLOCK_SIZE = 1024 * 4 // 4 KiB
+const MAX_BLOCK_COUNT = 50000
+const AZURE_BATCH_MAX_REQUEST = 256 // BATCH_MAX_REQUEST constant from Azure
+const { warn, info } = createLogger('xo:fs:azure')
+
+export default class AzureHandler extends RemoteHandlerAbstract {
+  #container
+  #dir
+  #blobServiceClient
+  #containerClient
+
+  constructor(remote, _opts) {
+    super(remote)
+    const { username, path, password, host, protocol } = parse(remote.url)
+    const credentials = new StorageSharedKeyCredential(username, password)
+    const url =
+      remote.protocol === 'http'
+        ? `${protocol}://${host}/${username}`
+        : `${protocol}://${username}.blob.core.windows.net`
+
+    info('Connecting to Azure blob storage...')
+    this.#blobServiceClient = new BlobServiceClient(url, credentials)
+
+    const parts = split(path)
+    this.#container = parts.shift() // in azurite, container = first component after host, only lowercase allowed
+    this.#dir = join(...parts)
+    this.#containerClient = this.#blobServiceClient.getContainerClient(this.#container)
+    const WITH_RETRY = ['_copy', '_getSize', '_list', '_outputFile', '_read', '_rename', '_unlink', '_writeFile']
+    WITH_RETRY.forEach(functionName => {
+      if (this[functionName] !== undefined) {
+        this[functionName] = pRetry.wrap(this[functionName], {
+          delays: [100, 200, 500, 1000, 2000],
+          when: err => !['SystemInUse'].includes(err?.code),
+          onRetry(error) {
+            warn('retrying method on fs ', {
+              method: functionName,
+              attemptNumber: this.attemptNumber,
+              delay: this.delay,
+              error,
+              file: this.arguments?.[0],
+            })
+          },
+        })
+      }
+    })
+  }
+
+  get type() {
+    return 'azure'
+  }
+
+  /**
+   *
+   * @param {string} path String to format
+   * @returns add / at the end of path to allow azure to understand it is a "folder"
+   */
+  #makePrefix(path) {
+    return path === '.' ? '' : path.endsWith('/') ? path : `${path}/`
+  }
+
+  #makeFullPath(path, isDir = false) {
+    let prefixedPath = path
+    if (path.startsWith('/')) {
+      prefixedPath = path.substring(1)
+    }
```

Comment:
IMO it looks better like this
```suggestion
    let prefixedPath = path.startsWith('/') ? path.substring(1) : path
```